go-controller/pkg/ovn/external_ids_syncer/port_group/port_group_sync.go

```diff
@@ -0,0 +1,230 @@
+	egressDefaultDenySuffix      = "egressDefaultDeny"
+	defaultNetworkControllerName = "default-network-controller"
+)
+
+type updatePortGroupInfo struct {
+	acls  []*nbdb.ACL
+	oldPG *nbdb.PortGroup
+	newPG *nbdb.PortGroup
+}
+
+type PortGroupSyncer struct {
+	nbClient libovsdbclient.Client
+	// txnBatchSize is used to control how many port groups will be updated with 1 db transaction.
+	// Set to 0 to disable batching
+	txnBatchSize int
+}
+
+func getControllerName(networkExternalID string) string {
+	if networkExternalID == "" {
+		return defaultNetworkControllerName
+	}
+	return networkExternalID + "-network-controller"
+}
+
+// NewPortGroupSyncer creates a PortGroupSyncer that will sync port groups without the new ExternalIDs for all
+// controllers. Controller name will be defined based on getControllerName() function.
+func NewPortGroupSyncer(nbClient libovsdbclient.Client) *PortGroupSyncer {
+	return &PortGroupSyncer{
+		nbClient:     nbClient,
+		txnBatchSize: 50,
+	}
+}
+
+func getPortGroupNamespaceDbIDs(namespace, networkExternalID string) *libovsdbops.DbObjectIDs {
+	controllerName := getControllerName(networkExternalID)
+	return libovsdbops.NewDbObjectIDs(libovsdbops.PortGroupNamespace, controllerName, map[libovsdbops.ExternalIDKey]string{
+		libovsdbops.ObjectNameKey: namespace,
+	})
+}
+
+func getPortGroupNetpolNamespaceDbIDs(namespace, direction, networkExternalID string) *libovsdbops.DbObjectIDs {
+	controllerName := getControllerName(networkExternalID)
+	return libovsdbops.NewDbObjectIDs(libovsdbops.PortGroupNetpolNamespace, controllerName, map[libovsdbops.ExternalIDKey]string{
+		libovsdbops.ObjectNameKey:      namespace,
+		libovsdbops.PolicyDirectionKey: direction,
+	})
+}
+
+func getPortGroupNetworkPolicyDbIDs(policyNamespace, policyName, networkExternalID string) *libovsdbops.DbObjectIDs {
+	controllerName := getControllerName(networkExternalID)
+	return libovsdbops.NewDbObjectIDs(libovsdbops.PortGroupNetworkPolicy, controllerName,
+		map[libovsdbops.ExternalIDKey]string{
+			libovsdbops.ObjectNameKey: fmt.Sprintf("%s_%s", policyNamespace, policyName),
+		})
+}
+
+func getPortGroupClusterDbIDs(baseName, networkExternalID string) *libovsdbops.DbObjectIDs {
+	controllerName := getControllerName(networkExternalID)
+	return libovsdbops.NewDbObjectIDs(libovsdbops.PortGroupCluster, controllerName, map[libovsdbops.ExternalIDKey]string{
+		libovsdbops.ObjectNameKey: baseName,
+	})
+}
+
+// getReferencingObjsAndNewDbIDs finds all object that reference stale port group and tries to create a new dbIDs
+// based on referencing objects
+func (syncer *PortGroupSyncer) getReferencingObjsAndNewDbIDs(oldHash, oldName, networkExternalID string) (acls []*nbdb.ACL,
+	dbIDs *libovsdbops.DbObjectIDs, err error) {
+	// find all referencing objects
+	aclPred := func(acl *nbdb.ACL) bool {
+		return strings.Contains(acl.Match, "@"+oldHash)
+	}
+	acls, err = libovsdbops.FindACLsWithPredicate(syncer.nbClient, aclPred)
```

Comment:
Would it be interesting to do this in a single pass performance wise instead of one pass per PG? You could have a set of PG hashes you are looking for, extract the hash with a regex and capture group, and crosscheck with that set. Something like that.

Reply:
I didn't invest time in this, because this code should only be run once in a cluster lifetime (port groups with new dbIDs are filtered out in the beginning), so didn't really try to optimize performance here, do you think it makes sense to improve?

Reply:
>  On the "search all acls at the same time": we can only query acls for the port group that will be updated in the next transaction, because transaction may change acl as it may reference port group updated in the previous transaction. So the best we can do is pre-evaluate a list of port groups for every batch, and list acls related to this subset, but I can't figure out a nice way to fins acls referencing a set of port groups (considering one acl may reference more than 1). Maybe you can help me here..

So I was thinking out of a single predicate run, you could end up with a map[pg]->acls. I guess at this time the ACLs and PGs don't need to be actual ACLs or PGs, just the necessary information to correlate them (uuids/hashes/whatever). You can use the client cache or other cache if necessary and/or defer building the actual information for the op when the actual op is built.

That leads me nicely to the weighted stuff you did. It's novel but unsure if it is more complicated that need be. I haven't take a detailed look at it, but I would just give the same weight to a PG op and an ACL op, so if you are updating a PG with 10 ACLs that would be a batch of 10, 11 or 12 depending on how you count ops. We already have the `BatchMap` function that you could feed with the previous suggested map. It almost does what we want but it does split the map entries and we might not want that here. So we can have a version of that that does not split.

Just ideas for you to consider. Might be missing something here.

Reply:
alright, building the pg to acl map was easier than I expected, so there is only 1 predicate search left.
The current definition of a weight includes port group and referenced acls update ops. I tried to count acl ops and pg update ops separately, but the difference between re-creating a port group with 5 acls and 5K acls is very large (not including the acls themselves, but just because of the port_group.ACLs list). So there is no way say how many port groups can be updated in 1 transaction, as it depends on the number of acls it has. Now when I am saying it, I am realizing that the number of ports may also affect transaction size, so I should probably include that to the port group weight too. lmk if that makes sense

Reply:
I had to run extra perf tests to account for ports (sigh) that didn't make the weight evaluation easier, but that is port groups nature, not me :P

Reply:
I thought that what added significant time cost was reflect going through field nesting while marshaling and not size of lists. For example, we didn't worry about the number of IPs in an address set before.

Aren't we bound to the same problem when the original PGs get added in the first place?

Reply:
we probably should've, but address sets a but less intense, as they only have a list of ips, and there shouldn't be many address sets with lots of ips, because they are shared. For netpols it has a higher chance to have both lots of ports (especially for ANP/BANP) and lots of acls (have seen 8K for egress firewall)
PG is update by adding new ports and acls with mutate mostly (even if not, 1 port group can probably always fit into 1 transaction), but here we don't update port group, we delete and create it instead, meaning we need to marshall all its fields at once.

Reply:
Yeah, I don't see how this wouldn't happen for AS or anything else in normal operation. So I was just after a potentially better generalization if we think the problem might be more prominent than just here, preferably based on ops. 

But anyway, lets tackle that when we actually have the problem I guess.